### DISCUSSION TITLE
[RFC] gather testing utils for ant-design-based UI library

### DISCUSSION THREAD

**Discussion by mortalYoung** [ant-design/ant-design#42032] [Ideas]
# Introduction

新增 @ant-design/testing 的包，导出相关函数为 Ant Design 的相关组件做测试

# Motivation
- 沉淀出基于 Ant Design 的组件库后，需要补充测试用例，不可避免的需要用到一些测试相关的工具函数。例如 openSelect 等方法。但是这些方法各不相同，且写起来不方便。

# Expected
期望能有一个库，把各个测试用到的方法做一个合集来解决上面遇到的问题。

## API
API 的设计考虑遵循的规范如下：
1. 各个方法按照**组件名称区分**，有利于**降低上手成本**，会用 Ant Design 就会用该库，例如：`input.fireXXXX`
2. 调用名**以 fire 开头，组件名中 onXXX 结尾**。例如：`input.fireChange` 即理解为 Input 组件中触发 `onChange` 函数的动作
3. 支持 configProvider 配置全局的变量，例如：`prefixCls` 的修改
4. 常用的通用函数，例如 `sleep` 可以放 `common` 下

## Usage
```tsx
// xxx.test.tsx
import { input } from '@ant-design/testing'

test('xxx',() => {
	const fn = jest.fn();
	const { container } = render(<CustomerInput onChange={fn} />)
	input.fireChange(container, 'test');
	expect(fn).toBeCalled();
})
```

# Related Collections

理论上来说，**每一个组件的 onXXX 方法貌似都可以有对应的 fireXXX 函数**。
考虑到部分组件实现起来比较简单，如 Model 的 onOpen 之类的，所以可是适当实现部分功能

【待补充，大致思路是如下】考虑抽出来的相关用到的函数如下：

## Basic
1. [async function sleep(timeout: number): Promise<void\>](https://github.com/react-component/dropdown/blob/master/tests/utils.js#L6-L12)

## Input
1. [function fireChange(container, value:string): void](https://github.com/react-component/input/blob/master/tests/index.test.tsx#L47)
2. [function fireClear(): void](https://github.com/react-component/input/blob/master/tests/index.test.tsx#L156)

## InputNumber
1. [function fireChange(container,value)](https://github.com/ant-design/ant-design/blob/HEAD/components/input-number/__tests__/index.test.tsx)
2. [function fireStep(container, upDown: boolean)](https://github.com/ant-design/ant-design/blob/HEAD/components/input-number/__tests__/index.test.tsx#L22-L32)

## DropDown
1. function fireOpen() // 实现起来貌似比较简单，先不考虑
2. [function fireClose()](https://github.com/react-component/dropdown/blob/master/tests/basic.test.tsx#L406)

## Select
1. [function fireOpen(container)](https://github.com/react-component/select/blob/master/tests/utils/common.ts#L13-L15)
2. [function fireSelect(container, index: number)](https://github.com/react-component/select/blob/master/tests/utils/common.ts#L17-L22)

## Cascader
1. [function fireOpen(container)](https://github.com/ant-design/ant-design/blob/HEAD/components/cascader/__tests__/index.test.tsx#L15-L18)
2. [function fireChange(container, menuIndex: number, itemIndex: number)](https://github.com/ant-design/ant-design/blob/HEAD/components/cascader/__tests__/index.test.tsx#L27-L36)

## DatePicker
1. [function fireOpen(container)](https://github.com/ant-design/ant-design/blob/HEAD/components/date-picker/__tests__/utils.ts#L4-L7)
2. [function fireChange(container, text)](https://github.com/ant-design/ant-design/blob/HEAD/components/date-picker/__tests__/utils.ts#L13-L27)
3. [function fireClose(container)](https://github.com/ant-design/ant-design/blob/HEAD/components/date-picker/__tests__/utils.ts#L13-L27)

**Comment by zombieJ** [comment]
这个不一定要做成官方的，因为内部其实也不会特地去维护一套。但是如果有兴趣的话可以发个三方包，官方可以推荐哈~

**Reply by afc163** [reply]
做出来了我们也能用哈

**Comment by Wxh16144** [comment]
> 如果有兴趣的话可以发个三方包

@mortalYoung 可以组织一下, 我也想参与贡献. (提个建议, RFC 里面链接使用 [commit-shas](https://docs.github.com/zh/get-started/writing-on-github/working-with-advanced-formatting/autolinked-references-and-urls#commit-shas), 对应页面按下键盘 `y` 键即可

**Comment by mortalYoung** [comment]
搞了一个仓库试试水：https://github.com/mortalYoung/ant-design-testing ，大部分的坑都没填，只实现了大致的思路

**Comment by xiaosansiji** [comment]
有可能做成框架无关的测试用例？这样社区实现的 vue 等技术栈可能也可以使用，这有助于帮助各技术栈版本实现，能尽量保持一致。
